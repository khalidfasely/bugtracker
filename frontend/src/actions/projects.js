import projectsApi from "../api/projects";

export const setProjects = (projects) => ({
    type: "SET_PROJECTS",
    projects
});

export const startSetProjects = () => {
    return (dispatch) => {
        return projectsApi().then((result) => {
            dispatch(setProjects(result.projects));
            return result;
        });
    };
};