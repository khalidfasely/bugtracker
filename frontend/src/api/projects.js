export default async () => {
    const res = await fetch('/api/projects');
    
    const data = await res.json();
    
    return data;
};