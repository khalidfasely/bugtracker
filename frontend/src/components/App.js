function App() {
  return (
    <div className="App">
      <header>
        Hello world from React        
      </header>
    </div>
  );
}

export default App;
