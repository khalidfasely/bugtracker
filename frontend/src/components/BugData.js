const BugData = ({ bug }) => (
    <div>
        <p>{bug?.id}</p>
        <p>{bug?.user?.username}</p>
        <p>{bug?.title}</p>
        <p>{bug?.description}</p>
        <p>{bug?.time}</p>
        <p>{bug?.classification}</p>
        <p>{`${bug?.active}`}</p>
        <div>
            {bug?.admins?.map(admin => <span key={admin} >{admin} </span>)}
        </div>
        <br />
        <div>
            {bug?.users_with?.map(user => <span key={user} >{user} </span>)}
        </div>
    </div>
);

export default BugData;