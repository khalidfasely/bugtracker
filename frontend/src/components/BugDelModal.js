import Modal from "react-modal";

const BugDelModal = ({ delModalOpen, setDelModalOpen, deleteBug }) => (
    <Modal
        isOpen={delModalOpen}
        onRequestClose={() => setDelModalOpen(false)}
    >
        If you delete this bug there is no way to return it!
        <button onClick={deleteBug} data-testid='delete_button'>Delete</button>
    </Modal>
);

export default BugDelModal;