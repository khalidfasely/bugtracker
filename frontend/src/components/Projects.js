import { Link, NavLink } from 'react-router-dom';
import { connect } from 'react-redux';
import { useEffect } from 'react';

const Projects = ({ uname, projects }) => {
    return (
        <div>

            {
                uname ?
                <div>
                {
                    projects.length > 0 ?
                    projects.map(project => (
                        <div key={project.id}>
                            <NavLink to={`/project/${project.id}`}>
                                <p>{project.name}</p>
                                <p>{project.time}</p>
                            </NavLink>
                        </div>
                    )) :
                    <div>No projects. Create one <Link to="/new">here</Link>.</div>
                }
                </div> :
                <div>
                    <Link to="/login">Login</Link> or <Link to="/register">Register</Link> to see your projects or create new ones!
                </div>
            }
        </div>
    );
};

const mapStateToProps = (state) => ({
    uname: state.auth.uname,
    projects: state.projects.projects
});

export default connect(mapStateToProps)(Projects);