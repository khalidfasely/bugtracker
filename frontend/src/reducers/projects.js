const projectsReducerDefaultState = {
    projects: []
}

export default (state = projectsReducerDefaultState, action) => {
    switch (action.type) {
        case 'SET_PROJECTS':
            return {
                ...state,
                projects: action.projects
            };
        
        case 'NEW_PROJECT':
            return {
                ...state,
                projects: [
                    action.project,
                    ...state.projects
                ]
            };
        
        case 'DELETE_PROJECT':
            return {
                ...state,
                projects: state.projects.filter(project => project.id !== action.pid)
            };

        default:
            return state;
    };
};