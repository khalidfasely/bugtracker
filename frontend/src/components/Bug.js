import { useState, useEffect } from "react";
import { connect } from "react-redux";
import { useParams } from "react-router-dom";
import { startSetBug } from "../actions/bug";

import BugItem from './BugItem';

export const Bug = ({ startSetBug }) => {
    const params = useParams();

    const [ableToSee, setAbleToSee] = useState(false);
    const [notFound, setNotFound] = useState('');

    useEffect(() => {
        startSetBug(params.pid, params.bid)
        .then((result) => {
            setNotFound('');

            if (result.message) {
                setNotFound(result.message);
            };
        })
        .then(() => setAbleToSee(true))
        .catch(er => console.error(er));
    }, [params])


    if (notFound) {
        return <div>Bug Not Found! -- {notFound}</div>;
    };

    return (
        <div>
            {
                ableToSee ?
                <div><BugItem /></div> :
                <div>You're not able to see this bug</div>
            }
        </div>
    );
};

const mapDispatchToProps = (dispatch) => ({
    startSetBug: (pid, bid) => dispatch(startSetBug(pid, bid))
});

export default connect(undefined, mapDispatchToProps)(Bug);