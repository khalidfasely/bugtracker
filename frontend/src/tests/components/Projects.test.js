import React from "react";
import { render } from "@testing-library/react";

import { Projects } from "../../components/Projects";
import projects from "../fixtures/projects";

let props;

beforeEach(() => {
    props = {
        uname: 'admin',
        projects: [projects[1]]
    }
});

test('Should render Projects component without props', () => {
    const { asFragment } = render(<Projects />);
    expect(asFragment()).toMatchSnapshot();
});

test('Should render Projects component with props', () => {
    const { asFragment } = render(<Projects {...props} />);
    expect(asFragment()).toMatchSnapshot();
});

test('Should render Projects component with projects empty array', () => {
    const { asFragment } = render(<Projects {...{...props, projects: []}} />);
    expect(asFragment()).toMatchSnapshot();
});