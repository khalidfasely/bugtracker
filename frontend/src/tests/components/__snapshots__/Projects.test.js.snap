// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Projects component with projects empty array 1`] = `
<DocumentFragment>
  <div>
    <div>
      <div
        class="projects__link-new"
      >
        No projects. Create one 
        <a>
          here
        </a>
        .
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Should render Projects component with props 1`] = `
<DocumentFragment>
  <div>
    <div>
      <div>
        <div
          class="projects__item"
        >
          <a
            class="nav-link"
          >
            <div
              class="projects__item-content"
            >
              <p
                class="projects__item-content__name"
              >
                <span
                  class="mobile-screen"
                >
                  project2...
                </span>
                <span
                  class="desktop-screen"
                >
                  project2...
                </span>
              </p>
              <div
                class="projects__item-content__us-ti"
              >
                <p
                  class="projects__item-content__user"
                >
                  admin
                </p>
                <p
                  class="projects__item-content__time"
                >
                  now
                </p>
              </div>
            </div>
          </a>
        </div>
        <div
          class="projects__link-new"
        >
          Create new project 
          <a>
            here
          </a>
          .
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Should render Projects component without props 1`] = `
<DocumentFragment>
  <div>
    <div
      class="projects__anonymous"
    >
      <a>
        Login
      </a>
       or 
      <a>
        Register
      </a>
       to see your projects or create new ones!
    </div>
  </div>
</DocumentFragment>
`;
